python/raydp/tests/test_data_owner_transfer.py

```diff
@@ -0,0 +1,174 @@
+
+import sys
+import os
+import time
+
+import pytest
+import ray
+import raydp
+
+
+def gen_test_data():
+  from pyspark.sql.session import SparkSession
+  s = SparkSession.getActiveSession()
+ 
+  data = [("ming", 20, 15552211521),
+          ("hong", 19, 13287994007),
+          ("dave", 21, 15552211523),
+          ("john", 40, 15322211523),
+          ("wong", 50, 15122211523)]
+
+  rdd = s.sparkContext.parallelize(data)
+  out = s.createDataFrame(rdd, ["Name", "Age", "Phone"])
+  return out
+  
+
+@pytest.mark.xfail(reason="data ownership transfer feature is not enabled")
+def test_fail_without_data_ownership_transfer():
+  """
+  Test shutting down Spark worker after data been put 
+  into Ray object store without data ownership transfer.
+  This test should be throw error of data inaccessible after
+  its owner (e.g. Spark JVM process) has terminated, which is expected.
+  """
+
+  from raydp.spark.dataset import spark_dataframe_to_ray_dataset
+  
+  num_executor = 4
+
+  ray.shutdown()
+  raydp.stop_spark()
+
+  ray.init()
+  spark = raydp.init_spark(
+    app_name = "example",
+    num_executors = num_executor,
+    executor_cores = 1,
+    executor_memory = "800M"
+    )
+
+  df_train = gen_test_data()
+  # df_train = df_train.sample(False, 0.001, 42)
+
+  resource_stats = ray.available_resources()
+  cpu_cnt = resource_stats['CPU']
+
+  # convert data from spark dataframe to ray dataset without data ownership transfer
+  df = spark_dataframe_to_ray_dataset(df_train, parallelism=4)
+
+  # display data
+  df.show(5)
+
+  # release resource by shutting down spark
+  raydp.stop_spark()
+  ray.internal.internal_api.global_gc() # ensure GC kicked in
+  time.sleep(3)
+
+  # confirm that resources has been recycled
+  resource_stats = ray.available_resources()
+  assert resource_stats['CPU'] == cpu_cnt + num_executor
+
+  # confirm that data get lost (error thrown)
+  df = df.to_dask()
+
+
+def test_data_ownership_transfer():
+  """
+  Test shutting down Spark worker after data been put 
+  into Ray object store with data ownership transfer.
+  This test should be able to execute till the end without crash as expected.
+  """
+
+  from raydp.spark.dataset import spark_dataframe_to_ray_dataset
+  import numpy as np
+  
+  num_executor = 4
+
+  ray.shutdown()
+  raydp.stop_spark()
+
+  ray.init()
+  spark = raydp.init_spark(
+    app_name = "example",
+    num_executors = num_executor,
+    executor_cores = 1,
+    executor_memory = "800M"
+    )
+
+  df_train = gen_test_data()
+
+  resource_stats = ray.available_resources()
+  cpu_cnt = resource_stats['CPU']
+
+  # convert data from spark dataframe to ray dataset,
+  # and transfer data ownership to dedicated Object Holder (Singleton)
+  df = spark_dataframe_to_ray_dataset(df_train, parallelism=4, _use_owner=True)
```

Comment:
can you call this variable ds, to align with other tests? ds is short for dataset, df is for dataframe

Reply:
done